### ISSUE TITLE
Consider subcommands for indexing

### ISSUE THREAD

**Issue by marcelm** [ksahlin/strobealign#59]
> I mainly want to ensure a good user interface. If a user says "only create an index" in some way, there should be an error if alignment options are used. They should not be silently ignored. If indexing and mapping functionality is split into two subcommands, this would be straightforward: The indexing subcommand accepts indexing parameters, and the mapping subcommand accepts lots of alignment parameters. So indexing options aren’t available when mapping and vice versa.

This is a solid point.  I guess the tradeoff is a good user interface vs an option to do a one command run (as for minimap2). Perhaps both can be achieved with the argparser we now use? Like `strobealign index` `strobealign align` or `strobealign idx-n-aln`?

> The question is what 'will not work' means at the moment. IMO, this must either not be possible (for example because there’s no way of providing index parameters when mapping) or at least throw a clear error message. It would be best if the required index parameters were just read from the index when a pre-generated one is used.

I meant it as in that there will be little to no correct alignments (but strobealign will likely still run without error). Most reads will probably be unmapped (depending on the magnitude of the difference in settings). @johan-gson pointed out that we could easily check for identical mapping parameters when we load the index and throw an error if they are different. I think that will be a nice addition.

_Originally posted by @ksahlin in https://github.com/ksahlin/StrobeAlign/issues/48#issuecomment-1231356627_

**Comment by marcelm** [comment]
Another quick suggestion for how subcommands could work:

1. Map without a pre-generated index:
```
strobealign map reference.fasta reads1 [reads2]
```

2. Create an index:
```
strobealign index [-r INT] reference.fasta [[reads1] reads2]
```
(Either `-r` must be used or at least one reads file provided.)

3. Map with a pre-generated index:
```
strobealign map -i reference.fasta [[reads1] reads2]
```
Note that this re-purposes `-i` to mean "use the index" instead of "create the index" (we should still have a `--use-index` alias).

This is a compromise: It would be cleanest to have a third command that does the `map -i` and that would not accept any of the indexing parameters, but I cannot come up with a nice name for it.

Or perhaps use `map` for "map with pre-generated index" and `directmap` or something like that for the combined indexing and mapping command?

**Comment by ksahlin** [comment]
I like this suggestion.

**Comment by marcelm** [comment]
I’m currently updating the section in the README describing how to use the `--create-index` and `--use-index` options. It’s getting a bit awkward formulating this because I think one should use a pre-generated index in most cases, and then it’s weird to have an extra flag for it. So I’m coming back to your last sentence here:

> Or perhaps use `map` for "map with pre-generated index" and `directmap` or something like that for the combined indexing and mapping command?

I now think this is the way to go: By default, make the mapping command *require* a pre-generated index to exist. Then we can have an extra subcommand (or flag) to force on-the-fly index generation. In fact, why not call it so:

```
# two-step indexing and mapping:
strobealign index -r 100 reference.fasta
strobealign map reference.fasta reads.fastq

# one-step indexing and mapping:
strobealign map --on-the-fly reference.fasta reads.fastq
```

**Comment by ksahlin** [comment]
> strobealign map --on-the-fly reference.fasta reads.fastq

I still believe the most common use case is on the fly mapping. Maybe I am wrong. But given this could we simply have `strobealign reference.fasta reads.fastq` (i.e. no sub command) for on the fly mapping? This looks the most canonical. Or do you think it would be confusing?

> strobealign index -r 100 reference.fasta

Will this command throw an error if one tries to do `strobealign index reference.fasta`?

**Comment by marcelm** [comment]
> I still believe the most common use case is on the fly mapping. Maybe I am wrong. But given this could we simply have `strobealign reference.fasta reads.fastq` (i.e. no sub command) for on the fly mapping? This looks the most canonical. Or do you think it would be confusing?

If you think that’s the most common use case, then the interface we have with `--create-index`/`--use-index` at the moment is fine. I suggest we leave this as it is for the moment and try to gauge what people do in practice by the feedback we get.

> > strobealign index -r 100 reference.fasta
> 
> Will this command throw an error if one tries to do `strobealign index reference.fasta`?

Implemented in #207 just before you posted this :smile:.

**Comment by marcelm** [comment]
I’m going to close this now as using subcommands sounds less and less interesting. `--create-index` and `--use-index` have become a little bit less relevant with the speed improvements in the last release and it may be good to not promote them from mere options to full subcommands. We want to emphasize the on-the-fly aspect.
